packages/binarytree/src/proof.ts

```diff
@@ -0,0 +1,51 @@
+import { equalsBytes } from '@ethereumjs/util'
+
+import { createBinaryTree } from './constructors.js'
+import { decodeBinaryNode } from './node/index.js'
+
+import type { BinaryTree } from './binaryTree.js'
+import type { BinaryNode } from './node/index.js'
+import type { StemBinaryNode } from './node/stemNode.js'
+
+/**
+ * Saves the nodes from a proof into the tree.
+ * @param proof
+ */
+export async function binaryTreeFromProof(_proof: Uint8Array[]): Promise<BinaryTree> {
+  const proofTrie = await createBinaryTree()
+  const putStack: [Uint8Array, BinaryNode][] = _proof.map((bytes) => {
+    const node = decodeBinaryNode(bytes)
+    return [proofTrie['merkelize'](node), node]
+  })
+  await proofTrie.saveStack(putStack)
+  const root = putStack[0][0]
+  proofTrie.root(root)
+  return proofTrie
+}
+
+/**
+ * Verifies a proof.
+ * @param rootHash
+ * @param key
+ * @param proof
+ * @throws If proof is found to be invalid.
+ * @returns The value from the key, or null if valid proof of non-existence.
+ */
+export async function verifyBinaryProof(
+  _rootHash: Uint8Array,
+  _key: Uint8Array,
+  _proof: Uint8Array[],
+): Promise<Uint8Array | null> {
+  const proofTrie = await binaryTreeFromProof(_proof)
+  const [value] = await proofTrie.get(_key.slice(0, 31), [_key[31]])
+  const valueNode = decodeBinaryNode(_proof[_proof.length - 1]) as StemBinaryNode
+  const expectedValue = valueNode.values[_key[31]]
+  if (!expectedValue) {
+    if (value) {
+      throw new Error('Proof is invalid')
```

Comment:
Would it make sense to give more detailed error messaging since we have different possible outcomes (i.e. `proof witness does not reach key` and `proof witness leads to different value` or something to that effect)?

Reply:
i think those are just two ways of saying the same thing.  `This Proof is invalid *for that key*`.

We don't actually verify that the provided proof is itself a valid path through a tree.  If we did, there could be a 'Bad Proof` error, but i'm not sure it's worth the CPU

Reply:
Good point.